File: clients/android/OrbitTermAndroid/data/src/main/java/com/orbitterm/android/sync/PortForwardProfileSyncContract.kt
package com.orbitterm.android.sync

import java.util.UUID
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject

@Serializable
internal data class PortForwardProfileSyncEnvelope(
    val kind: String,
    val version: Int,
    val updatedAtUnix: Long,
    val profiles: List<PortForwardProfileWire>,
    val tombstones: List<PortForwardProfileTombstoneWire>,
)

@Serializable
internal data class PortForwardProfileWire(
    val id: String,
    val assetId: String,
    val name: String,
    val mode: String,
    val bindHost: String,
    val bindPort: Int,
    val destinationHost: String,
    val destinationPort: Int,
    val createdAtUnix: Long,
    val updatedAtUnix: Long,
)

@Serializable
internal data class PortForwardProfileTombstoneWire(val id: String, val deletedAtUnix: Long)

@Serializable
internal data class PortForwardProfileVaultDocument(
    val version: Int = 1,
    val profiles: List<PortForwardProfileWire> = emptyList(),
    val localOnlyProfiles: List<PortForwardProfileWire> = emptyList(),
    val tombstones: Map<String, Long> = emptyMap(),
    val remoteConfigId: UInt? = null,
    val vectorClock: String = "{}",
    val payloadFingerprint: String = "",
)

/** Saved profile contract only; live tunnel/process/auto-start state is rejected. */
internal object PortForwardProfileSyncContract {
    const val MARKER = "orbit_port_forwards"
    const val VERSION = 1
    const val MAXIMUM_PROFILES = 256
    private const val MAXIMUM_RULES_PER_ASSET = 32
    private val forbiddenLiveFields = setOf(
        "tunnelId", "processId", "isRunning", "running", "autoStart", "startAfterVerifiedConnection",
    )
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }

    fun decode(plaintext: String): PortForwardProfileSyncEnvelope? = runCatching {
        val root = json.parseToJsonElement(plaintext).jsonObject
        require(!containsLiveState(root))
        validate(json.decodeFromJsonElement<PortForwardProfileSyncEnvelope>(root))
    }.getOrNull()

    fun encode(envelope: PortForwardProfileSyncEnvelope): String = json.encodeToString(validate(envelope))

    fun validate(envelope: PortForwardProfileSyncEnvelope): PortForwardProfileSyncEnvelope {
        require(envelope.kind == MARKER && envelope.version == VERSION && envelope.updatedAtUnix > 0)
        require(envelope.profiles.size <= MAXIMUM_PROFILES)
        require(envelope.tombstones.size <= MAXIMUM_PROFILES * 4)
        val profiles = envelope.profiles.map(::normalize)
        val tombstones = envelope.tombstones.map { item ->
            require(item.deletedAtUnix > 0)
            item.copy(id = canonicalUuid(item.id))
        }
        require(profiles.map(PortForwardProfileWire::id).distinct().size == profiles.size)
        require(tombstones.map(PortForwardProfileTombstoneWire::id).distinct().size == tombstones.size)
        require(profiles.groupingBy(PortForwardProfileWire::assetId).eachCount().values.all { it <= MAXIMUM_RULES_PER_ASSET })
        return envelope.copy(profiles = profiles, tombstones = tombstones)
    }

    private fun normalize(item: PortForwardProfileWire): PortForwardProfileWire {
        val name = normalizeText(item.name, 80)
        val bindHost = normalizeHost(item.bindHost)
        require(item.mode in setOf("local", "remote", "dynamicSocks5"))
        require(item.bindPort in 0..65535)
        require(item.createdAtUnix > 0 && item.updatedAtUnix >= item.createdAtUnix)
        val dynamic = item.mode == "dynamicSocks5"
        val destinationHost = if (dynamic) "" else normalizeHost(item.destinationHost)
        val destinationPort = if (dynamic) 0 else item.destinationPort.also { require(it in 1..65535) }
        return item.copy(
            id = canonicalUuid(item.id),
            assetId = canonicalUuid(item.assetId),
            name = name,
            bindHost = bindHost,
            destinationHost = destinationHost,
            destinationPort = destinationPort,
        )
    }

    private fun containsLiveState(root: JsonObject): Boolean {
        if (root.keys.any(forbiddenLiveFields::contains)) return true
        val profiles = root["profiles"]?.jsonArray ?: return false
        return profiles.any { element -> element.jsonObject.keys.any(forbiddenLiveFields::contains) }
    }

    private fun canonicalUuid(raw: String): String = UUID.fromString(raw.trim()).toString()

    private fun normalizeHost(raw: String): String = normalizeText(raw, 253).also { host ->
        require(host.none { it.isWhitespace() || it.isISOControl() })
    }

    private fun normalizeText(raw: String, maximum: Int): String = raw.trim().also { value ->
        require(value.isNotEmpty() && value.length <= maximum && value.none(Char::isISOControl))
    }
}

File: clients/windows/src/OrbitTerm.Application/Accounts/PortForwardProfileSyncContract.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using OrbitTerm.Application.Sessions;

namespace OrbitTerm.Application.Accounts;

internal sealed record PortForwardProfileSyncEnvelope(
    string Kind,
    int Version,
    long UpdatedAtUnix,
    PortForwardProfileWire[] Profiles,
    PortForwardProfileTombstoneWire[] Tombstones,
    [property: JsonIgnore] ulong RemoteId = 0,
    [property: JsonIgnore] string VectorClock = "{}");

internal sealed record PortForwardProfileWire(
    Guid Id,
    Guid AssetId,
    string Name,
    string Mode,
    string BindHost,
    int BindPort,
    string DestinationHost,
    int DestinationPort,
    long CreatedAtUnix,
    long UpdatedAtUnix);

internal sealed record PortForwardProfileTombstoneWire(Guid Id, long DeletedAtUnix);

/// <summary>
/// Versioned, non-secret profile contract shared by desktop and mobile clients.
/// It deliberately contains no tunnel handle, process identifier, running
/// state or auto-start flag. Receiving a profile can never start a tunnel.
/// </summary>
internal static class PortForwardProfileSyncContract
{
    internal const string Marker = "orbit_port_forwards";
    internal const int Version = 1;
    internal const int MaximumProfiles = 256;
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web);

    internal static PortForwardProfileSyncEnvelope? Parse(JsonElement root)
    {
        if (ContainsLiveState(root)) return null;
        PortForwardProfileSyncEnvelope? envelope;
        try { envelope = root.Deserialize<PortForwardProfileSyncEnvelope>(JsonOptions); }
        catch (JsonException) { return null; }
        if (envelope is null ||
            !string.Equals(envelope.Kind, Marker, StringComparison.Ordinal) ||
            envelope.Version != Version || envelope.UpdatedAtUnix <= 0 ||
            envelope.Profiles is null || envelope.Tombstones is null ||
            envelope.Profiles.Length > MaximumProfiles ||
            envelope.Tombstones.Length > MaximumProfiles * 4 ||
            envelope.Profiles.Select(item => item.Id).Distinct().Count() != envelope.Profiles.Length ||
            envelope.Tombstones.Select(item => item.Id).Distinct().Count() != envelope.Tombstones.Length)
        {
            return null;
        }

        var normalized = new List<PortForwardProfileWire>(envelope.Profiles.Length);
        foreach (var item in envelope.Profiles)
        {
            if (item.CreatedAtUnix <= 0 || item.UpdatedAtUnix < item.CreatedAtUnix ||
                !TryParseMode(item.Mode, out var mode)) return null;
            try
            {
                var rule = PortForwardingPolicy.Validate(new PortForwardingRule(
                    item.Id, item.AssetId, item.Name, mode, item.BindHost, item.BindPort,
                    item.DestinationHost, item.DestinationPort));
                normalized.Add(item with
                {
                    Name = rule.Name,
                    Mode = ToWireMode(rule.Mode),
                    BindHost = rule.BindHost,
                    BindPort = rule.BindPort,
                    DestinationHost = rule.DestinationHost,
                    DestinationPort = rule.DestinationPort,
                });
            }
            catch (ArgumentException) { return null; }
        }

        if (normalized.GroupBy(item => item.AssetId).Any(group => group.Count() > PortForwardingPolicy.MaximumRulesPerAsset) ||
            envelope.Tombstones.Any(item => item.Id == Guid.Empty || item.DeletedAtUnix <= 0)) return null;
        return envelope with { Profiles = normalized.ToArray() };
    }

    internal static byte[] Serialize(PortForwardProfileSyncEnvelope envelope)
    {
        var root = JsonSerializer.SerializeToElement(envelope, JsonOptions);
        var normalized = Parse(root) ?? throw new ArgumentException("端口映射配置同步信封无效。", nameof(envelope));
        return JsonSerializer.SerializeToUtf8Bytes(normalized, JsonOptions);
    }

    private static bool TryParseMode(string raw, out PortForwardingMode mode)
    {
        mode = raw switch
        {
            "local" => PortForwardingMode.Local,
            "remote" => PortForwardingMode.Remote,
            "dynamicSocks5" => PortForwardingMode.DynamicSocks5,
            _ => (PortForwardingMode)(-1),
        };
        return Enum.IsDefined(mode);
    }

    private static string ToWireMode(PortForwardingMode mode) => mode switch
    {
        PortForwardingMode.Local => "local",
        PortForwardingMode.Remote => "remote",
        PortForwardingMode.DynamicSocks5 => "dynamicSocks5",
        _ => throw new ArgumentOutOfRangeException(nameof(mode)),
    };

    private static bool ContainsLiveState(JsonElement root)
    {
        if (root.ValueKind != JsonValueKind.Object) return true;
        foreach (var forbidden in new[] { "tunnelId", "processId", "isRunning", "running", "autoStart" })
        {
            if (root.TryGetProperty(forbidden, out _)) return true;
        }
        if (!root.TryGetProperty("profiles", out var profiles) || profiles.ValueKind != JsonValueKind.Array) return false;
        foreach (var profile in profiles.EnumerateArray())
        {
            foreach (var forbidden in new[] { "tunnelId", "processId", "isRunning", "running", "autoStart", "startAfterVerifiedConnection" })
            {
                if (profile.ValueKind == JsonValueKind.Object && profile.TryGetProperty(forbidden, out _)) return true;
            }
        }
        return false;
    }
}
